closed/adds/jdk/src/windows/native/jdk/crypto/jniprovider/NativeCrypto_md.c

```diff
@@ -61,3 +87,108 @@ void * find_crypto_symbol(void *handle, const char *symname) {
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+/*
+ * Find path to JRE based on .exe's location
+ */
+jboolean
+GetJREPath(char *path, jint pathsize)
+{
+    char javadll[MAX_PATH];
+    struct stat s;
+    int rc;
+
+    if (GetApplicationHome(path, pathsize)) {
+        /* Is JRE co-located with the application? */
+        rc = JLI_Snprintf(javadll, sizeof(javadll), "%s\\bin\\" JAVA_DLL, path);
+        if (rc > 0 && rc <= MAX_PATH && (stat(javadll, &s) == 0)) {
+            return JNI_TRUE;
+        }
+        /* ensure storage for path + \jre + NULL */
+        if ((strlen(path) + 4 + 1) > pathsize) {
+            return JNI_FALSE;
+        }
+        /* Does this app ship a private JRE in <apphome>\jre directory? */
+        rc = JLI_Snprintf(javadll, sizeof (javadll), "%s\\jre\\bin\\" JAVA_DLL, path);
+        if (rc > 0 && rc <= MAX_PATH && (stat(javadll, &s) == 0)) {
+            strcat(path, "\\jre");
+            return JNI_TRUE;
+        }
+    }
+    return JNI_FALSE;
+}
+
+/*
+ * If app is "c:\foo\bin\javac", then put "c:\foo" into buf.
+ */
+jboolean
+GetApplicationHome(char *buf, jint bufsize)
+{
+    char *cp;
+    GetModuleFileName(0, buf, bufsize);
+    *strrchr(buf, '\\') = '\0'; /* remove .exe file name */
+    if ((cp = strrchr(buf, '\\')) == 0) {
+        /* This happens if the application is in a drive root, and
+         * there is no bin directory. */
+        buf[0] = '\0';
+        return JNI_FALSE;
+    }
+    *cp = '\0';  /* remove the bin\ part */
+    return JNI_TRUE;
+}
+
+/*
+ * windows snprintf does not guarantee a null terminator in the buffer,
+ * if the computed size is equal to or greater than the buffer size,
+ * as well as error conditions. This function guarantees a null terminator
+ * under all these conditions. An unreasonable buffer or size will return
+ * an error value. Under all other conditions this function will return the
+ * size of the bytes actually written minus the null terminator, similar
+ * to ansi snprintf api. Thus when calling this function the caller must
+ * ensure storage for the null terminator.
+ */
+int
+JLI_Snprintf(char* buffer, size_t size, const char* format, ...) {
+    int rc;
+    va_list vl;
+    if (size == 0 || buffer == NULL)
+        return -1;
+    buffer[0] = '\0';
+    va_start(vl, format);
+    rc = vsnprintf(buffer, size, format, vl);
+    va_end(vl);
+    /* force a null terminator, if something is amiss */
+    if (rc < 0) {
+        /* apply ansi semantics */
+        buffer[size - 1] = '\0';
+        return size;
```

Comment:
This doesn't seem right: I expect it should return `0` or `-1`.

Reply:
This is as-is copied from the existing Oracle code.

Reply:
The code you copied never has the return value checked until now: now that you are checking it, it should match expectations.

Reply:
Ah yes, now I take a good look at that code, if rc < 0 then an error has occurred and I don't see why it is doing buffer[size - 1] = '\0'; and then returning size. Think you are correct that this should just return -1;

Reply:
> an error has occurred and I don't see why it is doing buffer[size - 1] = '\0'; 

Maybe because of the documentation: "This function guarantees a null terminator under all these conditions.". Agree on the return values @keith-ca suggested.